### PR TITLE
Add replace option to random.categorical to enable sampling without replacement.

### LINKED ISSUE #25617: Add `replace: bool` argument to `random.categorical` to sample without replacement using Gumbel-top-k trick
[`random.categorical`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html) uses the Gumbel-max trick to sample with replacement.

An extension of the Gumbel-max trick allows sampling _without_ replacement. As noted in [this paper](https://arxiv.org/abs/1903.06059):

> The well-known Gumbel-Max trick for sampling from a categorical distribution can be extended to sample k elements without replacement.

> The Gumbel-Max trick [...] allows sampling from the categorical distribution, simply by _perturbing_ the log-probability for each category by adding _independent_ Gumbel distributed noise and returning the category with maximum _perturbed log-probability_. [...] However, there is more: as was noted (in a blog post) by Vieira (2014), taking the **top k largest perturbed log-probabilities** (instead of the maximum, or _top 1_) yields a sample of size k from the categorical distribution _without replacement_. We refer to this extension of the Gumbel-Max trick as the _Gumbel-Top-k_ trick.

**Proposal:**

Add a `replace: bool` argument to `random.categorical`. `replace=False` should use the Gumbel-Top-k trick (replacing [`argmax`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.argmax.html) with [`top_k`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.top_k.html)) to sample without replacement.

**Alternatives:**

[`random.choice`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.choice.html) can sample without replacement, but uses a less efficient [`cumsum`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.cumsum.html)-based method, and operates on probabilities rather than logits/log-probabilities. Related:
- https://github.com/jax-ml/jax/issues/22682
- https://github.com/jax-ml/jax/issues/25498

### PR DESCRIPTION
**Pull request by carlosgmartin**
Fixes https://github.com/jax-ml/jax/issues/25617.

### CODE REVIEW

File: tests/random_lax_test.py [hunk 1]
```diff
@@ -365,6 +366,39 @@ def testCategorical(self, p, axis, dtype, sample_shape):
...
+      assert output.shape == util.tuple_delete(logits_shape, axis, allow_neg=True)
+
+    else:
+      shape = batch_shape + util.tuple_delete(logits_shape, axis, allow_neg=True)
+
+      if math.prod(batch_shape) <= logits_shape[axis]:
+        output = random.categorical(key, logits, axis, shape, False)
+        assert output.shape == shape
+      else:
+        with self.assertRaises(Exception):
```
Comment (jakevdp) [inline tests/random_lax_test.py:398]: Use `assertRaisesRegex` with a more specific exception class and a regex for the expected error string.

File: jax/_src/random.py [hunk 2]
```diff
@@ -1548,12 +1549,44 @@ def _gumbel(key, shape, dtype, mode) -> Array:
...
+  operand = jnp.moveaxis(operand, axis, -1)
+  values, indices = lax.top_k(operand, k)
+  values = jnp.moveaxis(values, -1, axis)
+  indices = jnp.moveaxis(indices, -1, axis)
+  return values, indices
+
+
+def _categorical_without_replacement(
+  key: ArrayLike,
+  logits: RealArray,
```
Comment (jakevdp) [inline jax/_src/random.py:1563]: `key` and `logits` can be `Array` here I think, since they are cast to array by the calling function. Then you should be able to use the `.shape` attribute below.

File: jax/_src/random.py [hunk 3]
```diff
@@ -1548,12 +1549,44 @@ def _gumbel(key, shape, dtype, mode) -> Array:
...
+  shape: Shape | None = None,
+) -> Array:
+  if shape is None:
+    shape = util.tuple_delete(jnp.shape(logits), axis, allow_neg=True)
+  logits += gumbel(key, jnp.shape(logits))
+  batch_shape = shape[: len(shape) - jnp.ndim(logits) + 1]
+  batch_size = math.prod(batch_shape)
+  _, indices = _top_k(logits, batch_size, axis)
+  indices = jnp.moveaxis(indices, axis, 0)
+  indices = indices.reshape(shape)
```
Comment (jakevdp) [inline jax/_src/random.py:1574]: Minor: `lax.reshape` takes a `dimensions` permutation, so you should be able to do the permutation and the reshape in a single call.

That also suggests that calling `top_k` directly would be more efficient, because you wouldn't have to move the axis twice.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks pretty good! A few small comments below that I think will tighten-up the implementation a bit.

Thanks for putting this together!

File: jax/_src/random.py [hunk 4]
```diff
@@ -1562,15 +1595,25 @@ def categorical(key: ArrayLike,
...
+
+  References:
+    .. [1] Wouter Kool, Herke van Hoof, Max Welling. "Stochastic Beams and Where to Find
+      Them: The Gumbel-Top-k Trick for Sampling Sequences Without Replacement".
+      Proceedings of the 36th International Conference on Machine Learning, PMLR
+      97:3499-3508, 2019. https://proceedings.mlr.press/v97/kool19a.html.
   """
   key, _ = _check_prng_key("categorical", key)
   check_arraylike("categorical", logits)
   logits_arr = jnp.asarray(logits)
```
[the remark this answers is not in the document]
Comment (carlosgmartin) [reply jax/_src/random.py:1612]: The current implementation of `categorical` (with replacement) requires `axis` to be negative, hence [this line](https://github.com/jax-ml/jax/blob/aceae84fab5eb804111520fc3501e7a621d5e4ea/jax/_src/random.py#L1575). Would you like me to change it? (It's a bit convoluted.)

File: jax/_src/random.py (hunk 3 above)
Comment (carlosgmartin) [reply jax/_src/random.py:1574]: Is the `dimensions` argument of `lax.reshape` supposed to be the source indices, or the destination indices? The docs aren't clear about this. "the permutation order of the input shape" is ambiguous.

Comment (carlosgmartin) [reply jax/_src/random.py:1574]: Apparently it's the destination indices. I suggest rewording the docs to make that clearer.

**Comment by carlosgmartin** [comment]
@jakevdp I've edited `categorical` according to your comments to simplify both branches (with and without replacement).

File: jax/_src/random.py [hunk 5]
```diff
@@ -1562,32 +1569,46 @@ def categorical(key: ArrayLike,
...
-
-  batch_shape = tuple(np.delete(logits_arr.shape, axis))
+  logits = jnp.asarray(logits)
+  axis = canonicalize_axis(axis, logits.ndim)
   if shape is None:
-    shape = batch_shape
+    shape = util.tuple_delete(logits.shape, axis)
   else:
     shape = core.canonicalize_shape(shape)
-    _check_shape("categorical", shape, batch_shape)
```
Comment (jakevdp) [inline jax/_src/random.py:1582]: We should not remove this check – it validates that `shape` meets the documented requirements.

File: tests/random_lax_test.py [hunk 6]
```diff
@@ -365,6 +366,47 @@ def testCategorical(self, p, axis, dtype, sample_shape):
...
+  @jtu.sample_product(
+    [
+      {
+        "logits_shape": logits_shape,
+        "axis": axis,
+      }
+      for logits_shape in [(2,), (3,), (2, 3), (2, 5), (2, 3, 5)]
+      for axis in range(-len(logits_shape), len(logits_shape))
+    ],
+    batch_shape=[(), (2,), (3,), (2, 3), (2, 5), (2, 3, 5), None],
```
Comment (jakevdp) [inline tests/random_lax_test.py:378]: Maybe cleaner to leave out `None` here, as `()` covers that case – then you could remove the `if` block below.

File: tests/random_lax_test.py [hunk 7]
```diff
@@ -365,6 +366,47 @@ def testCategorical(self, p, axis, dtype, sample_shape):
...
+
+    key, subkey = random.split(key)
+    logits = random.normal(subkey, logits_shape)
+
+    if batch_shape is None:
+      output = random.categorical(key, logits, axis, None, False)
+      assert output.shape == util.tuple_delete(logits_shape, axis, allow_neg=True)
+
+    else:
+      shape = batch_shape + util.tuple_delete(logits_shape, axis, allow_neg=True)
```
Comment (jakevdp) [inline tests/random_lax_test.py:391]:
```suggestion
      shape = (*batch_shape, *np.delete(logits_shape, axis))
```

File: tests/random_lax_test.py [hunk 8]
```diff
@@ -365,6 +366,47 @@ def testCategorical(self, p, axis, dtype, sample_shape):
...
+
+        flat = output.reshape(batch_shape + (-1,))
+
+        # flat = output.reshape((-1,) + util.tuple_delete(logits.shape, axis, allow_neg=True))
+
+        assert jax.vmap(all_unique, [-1])(flat).all()
+      else:
+        with self.assertRaisesRegex(
+          ValueError,
+          "k argument to top_k must be no larger than minor dimension",
```
Comment (jakevdp) [inline tests/random_lax_test.py:405]: This is a pretty indirect error – could we detect this condition within `categorical` and raise something that's more actionable in that context?

File: jax/_src/util.py [hunk 9]
```diff
@@ -450,22 +450,24 @@ def wrapper(fun: T) -> T:
 def assert_unreachable(x):
   raise AssertionError(f"Unhandled case: {type(x).__name__}")
 
-def tuple_insert(t, idx, val):
```
Comment (jakevdp) [inline jax/_src/util.py:453]: I think once the other suggested changes are implemented, we could revert all changes to `jax/_src/util.py`.

File: jax/_src/random.py [hunk 10]
```diff
@@ -1562,32 +1569,46 @@ def categorical(key: ArrayLike,
     shape: Optional, a tuple of nonnegative integers representing the result shape.
       Must be broadcast-compatible with ``np.delete(logits.shape, axis)``.
       The default (None) produces a result shape equal to ``np.delete(logits.shape, axis)``.
+    replace: Perform sampling without replacement.
```
Comment (jakevdp) [inline jax/_src/random.py:1572]:
```suggestion
    replace: If True, perform sampling without replacement. Default (False) is to
      perform sampling with replacement.
```

File: jax/_src/random.py [hunk 11]
```diff
@@ -1562,32 +1569,46 @@ def categorical(key: ArrayLike,
...
+    logits += noise
+    return logits.argmax(batch_ndim + axis)
+
+  else:
+    logits += gumbel(key, logits.shape, logits.dtype)
+    k = math.prod(batch_shape)
+    logits = jnp.moveaxis(logits, axis, -1)
+    _, indices = lax.top_k(logits, k)
+
+    dimensions = (indices.ndim - 1,) + tuple(range(indices.ndim - 1))
```
Comment (jakevdp) [inline jax/_src/random.py:1609]:
```suggestion
    dimensions = (indices.ndim - 1, *range(indices.ndim - 1))
```

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/indexing.py [hunk 12]
```diff
@@ -397,7 +397,8 @@ def replace(tup, val):
 
 
 def _make_along_axis_idx(shape, indices, axis):
-  return tuple_replace(lax_numpy.indices(shape, sparse=True), axis, indices)
+  axis = canonicalize_axis(axis, len(shape))
+  return tuple_update(lax_numpy.indices(shape, sparse=True), axis, indices)
```
Comment (jakevdp) [inline jax/_src/numpy/indexing.py:401]: PRs should address a single thing: this cleanup in this file is unrelated to the main change in this PR, and should be reverted.

File: jax/_src/util.py [hunk 13]
```diff
@@ -462,10 +462,6 @@ def tuple_update(t, idx, val):
   assert 0 <= idx < len(t), (idx, len(t))
   return t[:idx] + (val,) + t[idx+1:]
 
-def tuple_replace(tupl, index, item):
-  # unlike tuple_update, works with negative indices as well
-  return tupl[:index] + (item,) + tupl[index:][1:]
-
```
Comment (jakevdp) [inline jax/_src/util.py:468]: Unrelated cleanup, please revert.

File: jax/_src/random.py [hunk 14]
```diff
@@ -1562,32 +1568,52 @@ def categorical(key: ArrayLike,
...
+    logits += noise
+    return logits.argmax(batch_ndim + axis)
+
+  else:
+    logits += gumbel(key, logits.shape, logits.dtype)
+    k = math.prod(batch_shape)
+    logits = jnp.moveaxis(logits, axis, -1)
+    try:
+      _, indices = lax.top_k(logits, k)
+    except Exception as e:
```
Comment (jakevdp) [inline jax/_src/random.py:1609]: We should not catch a base exception here: if `top_k` raises an unanticipated exception, this approach would obscure the true error.

I'd suggest that before calling `top_k`, we explicitly check the condition `k <= logits.shape[-1]`.

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Review by jakevdp — APPROVED** [review]

**Comment by carlosgmartin** [comment]
@jakevdp Looks like the changes triggered an error in a different test. I'm looking into it right now.

**Comment by carlosgmartin** [comment]
Just a comment, as an aside: The combined presence of the `logits_axis` and `result_shape` arguments (I disambiguate their names here), together with changing shapes, makes this trickier than I thought. That's why the original implementation for sampling with replacement was convoluted.

This is why I generally oppose arguments in the style of `result_shape`. It requires one to reason backwards from a desired outcome and solve a mental puzzle (higher chance of bugs) to get there. This contrasts with describing the generative process in a purely (straight)forward manner.

**Comment by jakevdp** [comment]
The semantics of `shape` here were a subject of a lot of discussion in the original contribution: #1855

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Hi - everything looks good, but merging is blocked by out-of-date files. Can you rebase against an updated main branch? Thanks!

**Comment by carlosgmartin** [comment]
@jakevdp Done.

**Review by jakevdp — APPROVED** [review]
